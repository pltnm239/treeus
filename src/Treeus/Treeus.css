.Treeus {
  --list-item-hover: rgba(0, 0, 0, 0.05);
  --list-item-focus: rgba(30, 167, 253, 0.07);
  --list-item-icon-transition: transform 0.1s ease-out;
  --list-item-icon-bg: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABmJLR0QA/wD/AP+gvaeTAAAAaUlEQVRIie2NMQqAMAwAb+oP3BScfVY3v2zBt7hEkKKtNqlTDgKBhDtwHEuWnq4AbEA0kEdxhfwwymFVyndgfnrQRKpyTeS1vCXyWX6NpEqkWX4yFSJqeSliJr+LmMvzSJK9C4OM4/zIAbQLGvIwlUVyAAAAAElFTkSuQmCC");

  width: 100%;
  margin: 0;
  padding: 0;
  list-style-type: none;
  font-size: 14px;

  box-sizing: border-box;
}

.Treeus__container {
  position: relative;
  min-height: 100%;
  overflow-x: hidden;
}

.Treeus__list {
  width: 100%;
  height: 100%;
  padding-left: 15px;
}

.Treeus__listItem {
  width: 100%;
  display: block;
  list-style-type: none;
  outline: none;
}

.Treeus__listItemLabel {
  display: block;
  padding: 6px 12px;
  border-radius: 5px;
  cursor: pointer;
}

.Treeus__listItemLabel:hover {
  background: var(--list-item-hover);
}

.Treeus__listItemLabel:before {
  display: inline-block;
  margin-right: 5px;
  background-image: var(--list-item-icon-bg);
  background-repeat: no-repeat;
  background-size: cover;
  width: 8px;
  height: 8px;
  position: relative;
}

.Treeus__listItem[data-has-children='true'] > .Treeus__listItemLabel:before {
  content: '';
  transform: rotate(-90deg);
  transition: var(--list-item-icon-transition);
}

.Treeus__listItem[data-open='true'] > .Treeus__listItemLabel:before {
  transform: rotate(0deg);
  transition: var(--list-item-icon-transition);
}

.Treeus__listItem[data-selected='true'] > .Treeus__listItemLabel {
  background: var(--list-item-focus);
}
