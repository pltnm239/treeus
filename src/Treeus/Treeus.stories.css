/* Dark theme */
.Treeus_dark {
    color: white;

    --list-item-focus: #393E46;
    --list-item-hover: rgba(181, 84, 0, 0.2);
    --list-item-icon-bg: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAyCAYAAAAeP4ixAAAABmJLR0QA/wD/AP+gvaeTAAAA20lEQVRoge3XOQ7CMBRF0WyCCPa/E6iYGlywnEuBXWChj5N4FO9KkVLE9j9ylWlSSimllFLqbwOOwA2YW88SB8zAHTilfHzhnesJ4xHOz3ZOWbADHn7BEzhUmHPpTPu1C5thViOMDapjNiOMjaphsiGMDYtjsiOMjYthiiGMA7JjiiOMg7JhqiGMAzdjqiOMg1djmiGMARZjmiOMQZIx3SCMgX5iukOElmC6RYRSMN0jQhZmGEToG2Y4RIjPvzkXvXfz15lUdAtj3UScv5mrf8a6CaWUUkoppbrtBWnjz9Xa11cNAAAAAElFTkSuQmCC");
}

.Treeus_dark .Treeus__listItem[data-has-children='true'] > .Treeus__listItemLabel:before {
    transform: rotate(-90deg);
}

.Treeus_dark .Treeus__listItem[data-open='true'] > .Treeus__listItemLabel:before {
    transform: rotate(0deg);
}

/* Custom render */
.Treeus_tennis .Treeus__listItemLabel:before {
    width: 0;
    background-image: none;
}
