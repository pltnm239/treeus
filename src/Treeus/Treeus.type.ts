import type { ReactNode, MouseEvent } from 'react';

type TOnClick = (event: MouseEvent, item: ITreeusItem) => void;

export interface ITreeusContext {
    selectedItemId: string;
    openItemsIds: string[];
    onListItemClick: TOnClick;
}

export interface ITreeusItem {
    id: string;
    label: string | ReactNode;
    children?: ITreeusItem[];
}

export interface ITreeusProps {
    items: ITreeusItem[];
    className?: string;
    onClick?: TOnClick;
    renderItemsCount?: number;
}
