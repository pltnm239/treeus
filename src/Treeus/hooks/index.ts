import useKeys from './useKeys';
import useListOptimization from './useListOptimization';
import useOpenItemsIds from './useOpenItemsIds';

export { useKeys, useListOptimization, useOpenItemsIds };
