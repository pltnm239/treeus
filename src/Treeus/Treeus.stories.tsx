import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Treeus } from './Treeus';
import { ITreeusProps, ITreeusItem } from './Treeus.type';
import './Treeus.stories.css';

export default {
    title: 'Treeus',
    component: Treeus,
    parameters: {
        options: { showPanel: true },
        backgrounds: {
            default: 'default',
            values: [
                { name: 'default', value: 'white' },
                { name: 'dark', value: '#222831' },
            ],
        },
    },
    decorators: [
        (Story) => (
            <div style={{ width: '150px' }}>
                <Story />
            </div>
        ),
    ],
} as ComponentMeta<typeof Treeus>;

const Template: ComponentStory<typeof Treeus> = (args) => <Treeus {...args} />;

export const Basic = Template.bind({});
Basic.args = {
    items: require('./test-data/mocks/Basic.json'),
};

export const DarkTheme = Template.bind({});
DarkTheme.parameters = {
    backgrounds: { default: 'dark' },
};
DarkTheme.args = {
    items: require('./test-data/mocks/DarkTheme.json'),
    className: 'Treeus_dark',
};

export const CustomRender = Template.bind({});
const getCustomRenderItems = () => {
    const tennisPlayers = [
        {
            name: 'Roger Federer',
            pic: 'https://lh3.googleusercontent.com/1dkVCzMCO5HoMT2PTZYldIBjbHpus6_lpcWytufSZAxRe2v7gehw2aI222fdoVH3oAEYMcn-_1udzzGPfnxC0D6dPDeETKMeU2dvMyw=s0',
        },
        {
            name: 'Rafael Nadal',
            pic: 'https://lh3.googleusercontent.com/drzSzbmVdZq5ZHNBm_YWhwPyHlIkQD8GdGxxUiRVRZqLGM-tx1m0NrCTylM2BIpe6wT2htCZ3afiIAflUu7eEoddkogc6ikn4z_HA54=s0',
        },
    ];

    let items = require('./test-data/mocks/CustomRender.json');
    items = items.map((item: ITreeusItem, index: number) => {
        item.label = (
            <span style={{ color: 'red' }}>
                <img width="12px" src={tennisPlayers[index].pic} />
                <span style={{ display: 'inline-block', marginLeft: '5px' }}>{tennisPlayers[index].name}</span>
            </span>
        );

        return item;
    });

    return items;
};
CustomRender.args = {
    items: getCustomRenderItems(),
    className: 'Treeus_tennis',
};

const longListStyle = {
    width: '250px',
    height: '570px',
    border: '1px solid rgba(0,0,0,0.1)',
    borderRadius: '5px',
};

export const ThousandElements = (args: ITreeusProps, { loaded: { items } }: { loaded: { items: ITreeusItem[] } }) => (
    <div style={longListStyle}>
        <Treeus {...args} items={items} />
    </div>
);

ThousandElements.loaders = [
    async () => ({
        items: require('./test-data/mocks/ThousandElements.json'),
    }),
];
ThousandElements.parameters = {
    options: { showPanel: false },
};

export const FiftyThousandElements = (
    args: ITreeusProps,
    { loaded: { items } }: { loaded: { items: ITreeusItem[] } },
) => (
    <div style={longListStyle}>
        <Treeus {...args} items={items} />
    </div>
);
FiftyThousandElements.parameters = {
    options: { showPanel: false },
};
FiftyThousandElements.loaders = [
    async () => ({
        items: require('./test-data/mocks/FiftyThousandElements.json'),
    }),
];

export const DifferentNesting = Template.bind({});
DifferentNesting.args = {
    items: require('./test-data/mocks/DifferentNesting.json'),
};
