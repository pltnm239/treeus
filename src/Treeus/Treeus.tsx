import React, { useState, memo, useRef, MouseEvent, useContext } from 'react';
import { useKeys, useListOptimization, useOpenItemsIds } from './hooks';
import type { ITreeusProps, ITreeusItem, ITreeusContext } from './Treeus.type';
import './Treeus.css';

const TreeusContext = React.createContext<ITreeusContext>({
    selectedItemId: '',
    openItemsIds: [],
    onListItemClick: () => null,
});

const TreeusItem: React.FC<ITreeusItem> = memo(({ id, label, children }) => {
    const { selectedItemId, openItemsIds, onListItemClick } = useContext(TreeusContext);
    const isOpen = openItemsIds.includes(id);
    const hasChildren = children && children.length > 0;
    const onClickEvent = (event: MouseEvent) => onListItemClick(event, { id, label, children });

    return (
        <li
            id={id}
            className="Treeus__listItem"
            data-selected={selectedItemId === id}
            data-open={isOpen}
            data-has-children={hasChildren}
            tabIndex={0}
        >
            <label className="Treeus__listItemLabel" onClick={onClickEvent}>
                {label}
            </label>
            {hasChildren && isOpen && (
                <ul className="Treeus__list">
                    {children.map((ch) => (
                        <TreeusItem {...ch} key={ch.id} />
                    ))}
                </ul>
            )}
        </li>
    );
});

export const Treeus: React.FC<ITreeusProps> = memo(({ items, onClick, className }) => {
    if (items?.length < 1) {
        return null;
    }

    let rootClassNames = 'Treeus';
    if (className) {
        rootClassNames += ` ${className}`;
    }

    const containerRef = useRef(null);
    const ulRef = useRef(null);
    const [renderItems, style] = useListOptimization(items, ulRef, containerRef);
    const [selectedItemId, setSelectedItemId] = useState<string>('');
    const [openItemsIds, toggleOpenItemIds, checkOpenItemId] = useOpenItemsIds([]);
    const [handleKeyDown, list, setIndex] = useKeys(
        items,
        setSelectedItemId,
        toggleOpenItemIds,
        checkOpenItemId,
        containerRef,
    );

    const onListItemClick = (event: MouseEvent, item: ITreeusItem) => {
        setSelectedItemId(item.id);
        toggleOpenItemIds(item.id);
        setIndex(list.findIndex((listItem) => listItem.id === item.id));

        if (onClick) {
            onClick(event, item);
        }
    };

    return (
        <TreeusContext.Provider value={{ selectedItemId, openItemsIds, onListItemClick }}>
            <div className="Treeus__container" ref={containerRef} tabIndex={-1}>
                <ul className={rootClassNames} ref={ulRef} style={style} onKeyDown={handleKeyDown}>
                    {renderItems.map((item) => (
                        <TreeusItem {...item} key={item.id} />
                    ))}
                </ul>
            </div>
        </TreeusContext.Provider>
    );
});
