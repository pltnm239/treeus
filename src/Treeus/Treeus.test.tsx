import React from 'react';
import { render } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import * as stories from './Treeus.stories';
const excludeTests = ['ThousandElements', 'FiftyThousandElements'];
const testCases = Object.values(composeStories(stories)).map((Story) => [Story.storyName!, Story]);

test.each(testCases)('Treeus %s', async (_storyName, Treeus) => {
    if (excludeTests.includes(_storyName as string)) {
        return;
    }

    const tree = await render(<Treeus />);
    const labels = document.querySelectorAll('Label');
    labels.forEach((label) => (label.innerHTML = 'test'));
    expect(tree.baseElement).toMatchSnapshot();
});
