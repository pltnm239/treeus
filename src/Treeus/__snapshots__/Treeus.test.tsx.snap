// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Treeus Basic 1`] = `
<body>
  <div>
    <div
      style="width: 150px;"
    >
      <div
        class="Treeus__container"
        tabindex="-1"
      >
        <ul
          class="Treeus"
        >
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="SU6mLp-z59wMvMC_ImHq_"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="9ESiSEAKTkfm4BwgR1lpX"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="_Z04PMu7kUV4cYeunCL0H"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="omeZRMgST-fHhWNI9t6vT"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="fVopM6JcICVmCeFJGAnBD"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`Treeus CustomRender 1`] = `
<body>
  <div>
    <div
      style="width: 150px;"
    >
      <div
        class="Treeus__container"
        tabindex="-1"
      >
        <ul
          class="Treeus Treeus_tennis"
        >
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="u2jVk7Ld8YO-6pDh3AG4C"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="xx8yuEfe68kqbtaWIfOJx"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`Treeus DarkTheme 1`] = `
<body>
  <div>
    <div
      style="width: 150px;"
    >
      <div
        class="Treeus__container"
        tabindex="-1"
      >
        <ul
          class="Treeus Treeus_dark"
        >
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="mNxvmMcEM63qwTzneBHde"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="QDPf6SoFikKDpuHAEEBG3"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="kO9U9xG_uOy0rm7wTytnN"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="J0dIFL-L9ACnMBvF5gad-"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="hBZrzWPKxlczIRmHzzE9D"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`Treeus DifferentNesting 1`] = `
<body>
  <div>
    <div
      style="width: 150px;"
    >
      <div
        class="Treeus__container"
        tabindex="-1"
      >
        <ul
          class="Treeus"
        >
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="Q7P1Fr-tDPnZtAWHHpKEC"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-open="false"
            data-selected="false"
            id="xtQ6ueYCjbdOkHowpIn9m"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
          <li
            class="Treeus__listItem"
            data-has-children="true"
            data-open="false"
            data-selected="false"
            id="7_FGVY6UY7SZfb2gXj5qS"
            tabindex="0"
          >
            <label
              class="Treeus__listItemLabel"
            >
              test
            </label>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;
